Adding new helper methods to the library

marcospont opened the pull request:
Hello!

A while back, I've wrote you an email about my intentions to offer some contributions to this package.

This PR contains the helpers that I had in my extensions of the library that I personally consider to be more helpful. They can actually offer reusable features other than just adding an extra layer on the call stack.

Hope you can review these changes and give your inputs about them.

I know that size matters and the original goal was to keep this library as short as possible. Hoping this extension proposal doesn't exceed the acceptable size threshold.

Thanks!

jquense commented on src/children.ts:
```diff
@@ -0,0 +1,7 @@
+export default function children(
+  node: Element | null
+): Element[] {
+  return collectSiblings(node ? node.firstElementChild : null);
```
what cases does this cover over `element.children` ?

marcospont replied:
In fact it does not add anything on top of the native.
On the other hand, the idea is to ensure the result child node list is an array.
The code can be changed to call splice on the children property instead of calling collectSiblings.

jquense replied:
what about `Array.from(node.children)`?

marcospont replied:
Sounds good, since it's way smaller

jquense commented on src/clear.ts:
```diff
@@ -0,0 +1,11 @@
+export default function clear(
+  node: Node | null
+): Node | null {
+  if (node) {
```
`node.innerHTML =''` might be more efficient?

marcospont replied:
I can see some recent benchmarks stating that removeChild loop is better than innerHTML.
Furthermore, in order to keep supporting IE11 for a while, removeChild loop is cross browser compliant because you can't set innerHTML of a <select> element to an empty string.

https://www.measurethat.net/Benchmarks/Show/6910/0/innerhtml-vs-removechild-vs-remove

jquense commented on src/isVisible.ts:
```diff
@@ -0,0 +1,5 @@
+export default function isVisible(
+  node: HTMLElement | null
+) : boolean {
+  return node ? !!(node.offsetWidth || node.offsetHeight || node.getClientRects().length) : false;
```
do you not need to check if `display` is something other than `none`? or opacity and visibility for that matter

marcospont replied:
This code was extracted from the code of the :visible pseudo-selector implementation in jQuery, even in the latest release 3.6.0. I'd say that testing for style display is not part of it because it would require to use getComputedStyle which can possibly be slower than reading the element's dimensions. I can look for benchmarks on this topic as well.

jquense replied:
I think this is fine, It's more that "visible" here is a bit of a misnomer but still a useful helper as is

jquense reviewed (commented):
looks good minus a few questions. Can you also add doc comments for the new methods as a form of documentation

marcospont reviewed (commented):
Hey @jquense can u elaborate more on what you mean about these doc comments? Should I follow any standards, are there examples I can check?

jquense commented:
yeah sorry. Comments like these: https://github.com/react-restart/hooks/blob/master/src/usePrevious.ts basically just a quick description on what it does and/or how to use it. These comments show up in intellisense overlays in text editors like vscode. I guess technically the standard is https://tsdoc.org/ but it's not important to be very thorough or use all the tags. It's basically jsdoc light

marcospont commented:
Hey, @jquense

I have added doc comments to all the new methods and also for most of the existing nodes (except the ones I wasn't secure on how to document them). Also, I have renamed `children` to `childElements` and added a new helper to get or set element attributes.

Can u please review it again? Tks.
